### PR TITLE
code review documentation

### PR DESCRIPTION
**Pull request by ericbuckley**
# Pull Request

## Description
This PR aims to initiate a discussion to standardize our code review process, with the intention of incorporating agreed-upon artifacts to provide guidance for the team in the future.

## Additional Notes
The content herein is not meant to be definitive or anywhere near our team's desired outcome. Rather, it serves as an initial framework, incorporating widely-used boilerplate definitions available online, along with some insights drawn from my past team experiences.

**If anyone finds this initial framework too opinionated or inflexible, please let your lead(s) know.  Starting over is always an option.**

The documents within are as follows:

- **.github/pull_request_template.md**: A checklist template that will be used for all future pull requests, aiming to maintain consistency for both author(s) and reviewer(s).
- **REVIEW_GUIDELINES.md**: A collection of guidelines, emphasizing flexibility rather than rules, outlining how reviews should be conducted. While the first rule, "Be Respectful," is non-negotiable, the rest are meant to be considered as guiding principles rather than inflexible mandates. Exceptions may apply to most of these guidelines at certain times.
- **REVIEW_SLAS.md**: A collection of anticipated procedures detailing the typical review process and agreements our team seeks to establish, fostering clear expectations between authors and reviewers.

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [ ] I have added or updated test cases to cover my changes, if applicable.
- [x] I have minimized the number of reviewers to include only those essential for the review.
- [ ] I have notified teammates in the review thread to build awareness.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

cc @zdeveloper @jonchang @colekettler @arinkulshi-skylight @knguyenrise8 @bryanbritten @alhayward @cbrinson-rise8 @derekadombek

### CODE REVIEW

**Comment by ericbuckley** [comment]
@jonchang @colekettler @arinkulshi-skylight @knguyenrise8 @alhayward @cbrinson-rise8 @derekadombek even though I didn't explicitly tag you as reviewers in this PR, I highly value your input and ideas.  Therefore, if you have the time, I encourage you to read these documents and share your thoughts.

**Review by bryanbritten — APPROVED** [review]
Thank you for putting this together! This all looks good to me. The proposal seems to introduce little to no new friction for developers, and seems like it will help improve the efficiency of code reviews.

**Comment by alhayward** [comment]
Looks great! Thanks for putting this together!

File: REVIEW_SLAS.md
Comment (derekadombek) [inline REVIEW_SLAS.md:1]: This is an interesting approach to code reviews and something I've honestly never thought about using an SLA on before. Not opposed to it at all because I've seen this effectively work for customer facing expectations but are there team ramifications for if the metrics are too high? or is it just something to measure and constant improve one?

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+## Related Issues
+[Link any related issues or tasks from your project management system.]
+
+## Additional Notes
+[Add any additional context or notes that reviewers should know about.]
+
+## Checklist
+Please review and complete the following checklist before submitting your pull request:
+
+- [ ] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
```
Comment (cbrinson-rise8) [inline .github/pull_request_template.md:15]: I like this idea and am wondering how the developer would ensure that the PR is a manageable size. And what the developer should do if the changes in a PR get too big. (curious because I normally don't think about the size of a PR as long as I'm fulfilling the related issue ticket)

File: REVIEW_SLAS.md
Comment (ericbuckley) [reply REVIEW_SLAS.md:1]: In most teams I've been a part of, agreements naturally develop as we become familiar with each other's habits. (occasionally I've encountered the engineer who won't review code, but that usually stems from other issues).  However, the purpose of this document is to establish clear expectations from the outset. By doing so, we can ensure that everyone is aligned, and we don't have to wait many months to understand each other's expectations and preferences.

Honestly, this could apply to a number of communication forms in asynchronous workflows.  Take Slack, for instance.  Some of us might think that waiting 48 hours to respond to a DM is acceptable, but others may feel obligated to respond ASAP.  The truth is, there is no expectation until we discuss and agree upon it.

This document is just trying to setup some expectations across the team.  If the team has consistent problems hitting these marks, then the team either needs to re-norm or managers/leads need to step in to have conversations.

@bryanbritten is there anything you'd add?

File: .github/pull_request_template.md (hunk 1 above)
Comment (ericbuckley) [reply .github/pull_request_template.md:15]: It could stand to have a metric attached to it for clarity, but I struggle to find a good one.

LoC changed could work, as I'm sure many of us would be unhappy to see a 10k LoC changed pull request to review.  However, there may be times when this makes sense.  Take a large project, with a simple refactor, many of those can easily hit the thousands of lines changed, but the amount of mental energy necessary to review is low.  So there are exceptions.

Hours, might be closer to what we want, but each of us likely review at different speeds.

I've intentionally left this open-ended to encourage the author(s) to approach it as a genuine question of good faith. Have you made an effort to break down this problem into manageable components, ensuring that the reviewers won't need days or weeks to comprehend the changes being made?

With all that said, if you have a suggestion to re-word, I'm all ears.

Comment (cbrinson-rise8) [reply .github/pull_request_template.md:15]: I'm fine with the wording. Especially if its meant to be open ended. I like it as a check box to think about when you're doing a PR. Thanks!
